Objects/obmalloc.c

```diff
@@ -1477,18 +1477,20 @@ _PyMem_DebugRawRealloc(void *ctx, void *p, size_t nbytes)
         /* overflow:  can't represent total as a Py_ssize_t */
         return NULL;
 
+    if (nbytes < original_nbytes) {
+        /* shrinking:  mark old extra memory dead */
+        memset(q + nbytes, DEADBYTE, original_nbytes - nbytes + 2*SST);
+    }
     /* Resize and add decorations. We may get a new pointer here, in which
      * case we didn't get the chance to mark the old memory with DEADBYTE,
      * but we live with that.
      */
-    oldq = q;
     q = (uint8_t *)api->alloc.realloc(api->alloc.ctx, q - 2*SST, total);
-    if (q == NULL)
+    if (q == NULL) {
+        if (nbytes <= original_nbytes) {
+            Py_FatalError("Shrinking reallocation is failed");
```

Comment:
Why the fatal error? If a shrinking `realloc()` fails, the old memory is still valid (just too large).

Reply:
But why NULL is returned instead of the old memory? This looks as a bug in the underlying allocator. I think a load crash is better than silent memory leak.

Reply:
Hmm, I have to think about this more. I'm confused by the recursion and I don't know what the intended behavior of api.realloc() is.

Reply:
Looking at the PEP example:

```void* my_realloc(void *ctx, void *ptr, size_t new_size)
{
    int padding = *(int *)ctx;
    return realloc(ptr, new_size + padding);
}
```

returning NULL and leaving the existing memory as is in the case of shrinking looks
correct to me.  That's how `realloc()` itself behaves.


[still confused by the recursion, but that seems a separate issue.]

Reply:
Please don't call Py_FatalError() in a memory allocator.